src/irr/asset/CGLILoader.cpp

```diff
@@ -0,0 +1,457 @@
+			imageInfo.arrayLayers = texture.faces() * texture.layers();
+
+			auto image = ICPUImage::create(std::move(imageInfo));
+
+			const auto texelBlockByteSize = asset::getTexelOrBlockBytesize(imageInfo.format);
+			auto texelBuffer = core::make_smart_refctd_ptr<ICPUBuffer>(texture.size());
+			auto data = reinterpret_cast<uint8_t*>(texelBuffer->getPointer());
+
+			auto regions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<ICPUImage::SBufferCopy>>(imageInfo.mipLevels);
+
+			auto getFullSizeOfRegion = [&](const uint16_t mipLevel) -> uint64_t
+			{
+				return texture.size(mipLevel) * imageInfo.arrayLayers;
+			};
+
+			auto getFullSizeOfLayer = [&](const uint16_t mipLevel) -> uint64_t
+			{
+				return texture.size(mipLevel);
+			};
+
+			{
+				uint16_t regionIndex = {};
+				uint64_t offset = {};
+				for (auto region = regions->begin(); region != regions->end(); ++region)
+				{
+					region->imageExtent.width = texture.extent(regionIndex).x;
+					region->imageExtent.height = texture.extent(regionIndex).y;
+					region->imageExtent.depth = texture.extent(regionIndex).z;
+					region->imageSubresource.mipLevel = regionIndex;
+					region->imageSubresource.layerCount = imageInfo.arrayLayers;
+					region->imageSubresource.baseArrayLayer = 0;
+					region->bufferOffset = offset;
+
+					offset += getFullSizeOfRegion(regionIndex);
+					++regionIndex;
+				}
+			}
+
+			uint64_t tmpDataSizePerRegionSum = {};
+			for (uint16_t mipLevel = 0; mipLevel < imageInfo.mipLevels; ++mipLevel)
+			{
+				const auto layerSize = getFullSizeOfLayer(mipLevel);
+				for (uint16_t layer = 0; layer < imageInfo.arrayLayers; ++layer)
+				{
+					const uint16_t gliLayer = texture.layers() > 1 ? layer / 6 : 0;
+					const uint16_t gliFace = texture.faces() > 1 ? layer % 6 : 0;
+
+					assignGLIDataToRegion((reinterpret_cast<uint8_t*>(data) + tmpDataSizePerRegionSum + (layer * layerSize)), texture, gliLayer, gliFace, mipLevel, layerSize);
+				}
+				tmpDataSizePerRegionSum += getFullSizeOfRegion(mipLevel);
+			}
+
+			image->setBufferAndRegions(std::move(texelBuffer), regions);
+
+			ICPUImageView::SCreationParams imageViewInfo;
+			imageViewInfo.image = std::move(image);
+			imageViewInfo.format = format.first;
+			imageViewInfo.viewType = imageViewType;
+			imageViewInfo.components = format.second;
+			imageViewInfo.flags = static_cast<ICPUImageView::E_CREATE_FLAGS>(0u);
+			imageViewInfo.subresourceRange.baseArrayLayer = 0u;
+			imageViewInfo.subresourceRange.baseMipLevel = 0u;
+			imageViewInfo.subresourceRange.layerCount = imageInfo.arrayLayers;
+			imageViewInfo.subresourceRange.levelCount = imageInfo.mipLevels;
+
+			auto imageView = ICPUImageView::create(std::move(imageViewInfo));
+
+			return SAssetBundle({std::move(imageView)});
+		}
+
+		bool CGLILoader::isALoadableFileFormat(io::IReadFile* _file) const
+		{
+			return true; // gli provides a function to load files, but we can check files' signature actually if needed
```

Comment:
check that stuff, otherwise the system will always proceed to load and fail files that are not of GLI loadable type.

Reply:
to do